BUG: Proposed bugfix/enhancement for unix platform (static/dynamic linked libstdc as an option).

oprienrico opened the pull request:
## Added new script to run extra configurations that are specific to the platform currently in use (loaded at the init of lead.m)
here it can handle extra path requirements, or other platform related customizations, such as the exclusion (which is the new default) or inclusion of the package libstdc++.so.6 for unix systems.
```
    root/helpers/ea_run_extra_machinespecific_init.m
```
which reads from the preferences set as prefs.platform.*  in the prefs file (and updated the defaults):
```
    root/common/ea_prefs_default.m
```
For this library (libstdc++.so.6) the preferred action is to use the toolchain that is already available in the system (e.g. using packaged build-essentials), for linux. Preferrably, also install the package matlab-support (on linux/ubuntu), and choose to use system libraries (and renaming the matlab ones).
This is set by the new preference line:
```
    prefs.platform.unix.load_shipped_libstdcpp6=false;
```

ningfei commented:
Hi Enrico, thanks for the PR. I noticed that the `libstdc++.so.6` shipped with LeadDBS might induce some problem if it's not updated. For the new proposed option `prefs.platform.unix.load_shipped_libstdcpp6=false;`, however, I'm not sure if it will have some side effects if it's set to `true`. Currently the c++ lib is only added to search path (for Linux, it's `LD_LIBRARY_PATH`) using the function`ea_libs_helper` when needed, for example, for some binaries like ANTs, SurfIce or BRAINSFit. I've actually been considering to to remove it from the path after it's used, to avoid messing up the c++ lib MATLAB uses on a global level. If we have such an option `load_shipped_libstdcpp6` set to true, with which LeadDBS modifies MATLAB search path on opening, do you think it might be a problem for other toolboxes in some cases? But I agree with you that it's better to install MATLAB support and rename the sys libs.

oprienrico commented:
Hi @ningfei , you are definitely right, the currently suggested bugfix could create an issue when it is set to true for the other tools. In addition I did not fully address correctly issue at hand. (I've revised with a commit, details below).
Regarding the `addpath()`, before this pull request, the `libstdc++.so.6` would have been added to the path and `LD_LIBRARY_PATH` in all cases, which caused the aforementioned issue. The `addpath()` or `rmpath()` were not the "culprits", so I have removed that part.
Specifically it errors out/conflict (in terminal) with the following error : 
```
/usr/local/MATLAB/R2020b/bin/glnxa64/matlab_helper: /home/user/develop/common/leaddbs_dev/ext_libs/support/unix/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by /usr/local/MATLAB/R2020b/bin/glnxa64/matlab_helper)
```
I believe this happens while executing :Chris Rorden's dcm2nii :: 2MAY2016 64bit BSD License 

However, after your feedback, I've addressed the `libstdc++.so.6` issue by modifying `ea_libs_helper` instead (eliminating the `ea_run_extra_machinespecific_init.m` and its reference from `lead.m`)
In addition, the `root/ext_libs/support folder` now includes three subfolders, each referred to the respective system (`pcwin`, `mac`, `unix`).
To allow to have an unified ea_prefs, I've currently left three platform suboptions, like this:
```
prefs.platform.unix.load_shipped_libstdcpp6=false;  %for UNIX/LINUX default is NOT loaded (using system libs)
prefs.platform.mac.load_shipped_libstdcpp6=true;    %for MAC default is loaded
prefs.platform.pcwin.load_shipped_libstdcpp6=false;  %for PC/WINDOWS default is NOT loaded
```
In this way it would follow the current planned behavior/call of `ea_libs_helper`, without cluttering the PATH handling with an extra script.
Let me know if this would work!

oprienrico commented on ext_libs/support/ea_libs_helper.m:
```diff
@@ -25,7 +51,11 @@ function ea_libs_helper(libpath, setpath)
+            if ispc()
+                setenv(envname, [libpath, ';', env]);%for pc is ";"
+            else
+                setenv(envname, [libpath, ':', env]);%for unix (including mac, is ":")
```
just changed to reflect the system specific separator (for the ENV variables)

oprienrico commented on ext_libs/support/ea_fix_runtimepath.m:
```diff
@@ -20,7 +20,11 @@
+        if ispc()
+            env = [env, ';', runtimePath{i}];%for pc is ";"
+        else
+            env = [env, ':', runtimePath{i}];%for unix (including mac, is ":")
```
just changed to reflect the system specific separator (for the ENV variables)

ningfei commented:
Looks good! Thanks :) I'm going to merge it now. For the dcm2niix problem, I guess I might compiled the current binary in host system instead of in docker. Will also update it soon. Btw, libstdc++ is only needed for Linux I think. For windows and mac, it's not needed. On mac, however, there's another issue about libSystem.B.dylib. Will also fix it the the upcoming release.

oprienrico commented:
Ok! This helps as i actually had issues with this library on linux. 
This hotfix can be removed in the future, when there is no more dependency issues. Or it could be left to enable users to enable/disable external libs as needed (I can only test on win and linux, as i do not have a mac).

ningfei commented:
Yes, I saw exactly the same error as you showed when I updated to MATLAB R2020b. Now I updated the shipped libstdc++. Let's see if it works better. Thanks again for your PRs!!
